---
title: "Informatics Challenge! 2022"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# How the informatics challenge will work

Welcome to the 2022 annual informatics challenge!

Below are a series of questions of varying difficulty. Each question will be
worth 1 - 5 points (denoted in the section header) before the questions. Some of
these questions are language specific, while others can be completed however
you'd like. Either way, **please show your work/code**. I've tried to specify in
each header if the questions are language specific, but feel free to ask for
clarification if needed.

I'm writing the questions to be either a task or specific question, such that
it is unambiguous if the answer is right. If you do the task or answer the
question correctly you get full points. No partial credit.

I am asking participants to specify their experience level so I can split folks
into two categories. I'll be awarding two trophies, one for each category.

-  Beginners I'm generally defining as those that are just getting started and have
been doing informatics regularly for a year or less.
-  You should likely submit as Advanced if you have been doing this for a while
and if all the beginner questions seem really easy and the harder questions
don't seem too bad.

There are several questions, so I'm giving two weeks to complete everything. You
don't have to complete all the questions, I'll be defining winners of the two
categories based on the highest total points. If you're the only one who enters,
you can win by answering just one question!

You can put your answers in-line in the Rmd file or you can provide your answers
another way if you prefer.

**The due date will be April 28th at 11:59 PM**

Feel free to ask for clarification on anything!

------------------------------------------------------------------------

## Are you submitting in the beginner or advanced category?

#Beginner
------------------------------------------------------------------------

# Beginner questions - 1 point per question

## --R-specific-- Data and variable types

1.  Make a numeric vector at least three elements long
```{r}
my_num <- c(1:3)
print(my_num)
```

2.  Make a named list object where the values are 10, 15, 20 and 25 and the
names of the elements are "Billy", "Eli", "Ali" and "Jordan"
```{r}
my_list <- list(Billy = 10,
                Eli = 15,
                Ali = 20,
                Jordan = 25)
print(my_list)
```

3.  Make a data frame where one column is random foods and the second column is
made-up calorie values

```{r}
my_df <- data.frame(
  Random_food = c ("Biryani", "Karahee", "Pakoray"),
  Calories = c(300, 400, 350)
  ) 
print(my_df)
```
4.  Add a new column of some sort to the data frame you made in question 3 and
assign the output a new variable
```{r}
new_df <- cbind(my_df, Origin = c("Karachi", "Faisalabad", "Lahore"))
print(new_df)
```

5.  Remove the column you just added and assign this to a new variable
```{r}
cities <- new_df$Origin
print(cities)
```
## Make basic plots

You can access a test dataset in R using the following code:

```{r}
mtcars <- mtcars
```

If you want to use a language other than R, the data can be downloaded here: https://nationwidechildrens-my.sharepoint.com/:t:/g/personal/matthew_cannon_nationwidechildrens_org/EbYpZRiuizFOjd9nz4vKn38Br8z2hEKnRijltqbzsUgnow?e=pPnNxW

6.  Using the mtcars data, make a scatterplot of displacement (disp) vs mpg
```{r}
attach(mtcars)
#MVC############################################################################
## To keep things tidy, if a function has > 2 arguments, put each on their own line
## Also the tidyverse style guide suggests spaces before/after "=" https://style.tidyverse.org/
plot(disp,
     mpg,
     main = "Scatterplot Example",
     xlab = "Displacement",
     ylab = "Miles Per Gallon ",
     pch = 15)
```

7.  Using the mtcars data again, make a boxplot of hp separated by cyl

```{r}
library(ggplot2)
mtcars %>% 
  ggplot(aes(x=factor(cyl), hp)) +
  geom_boxplot(fill = "steelblue") +
  ggtitle("Box Plot") +
  xlab("Cylinder") +
  ylab("hp")+
  theme_classic()
```
8.  Using the mtcars data again, make a barplot of displacement (disp) for each
engine shape (column "vs")
```{r}
q8 <- mtcars %>%
  ggplot(aes(x=factor(vs), disp)) + 
  geom_bar(stat = "identity", width = 0.6, fill = "steelblue") +
  ggtitle("Bar Plot") +
  xlab("Each engine shape") + ylab("Displacement") +
  theme_bw()
#MVC############################################################################
# unless you're in a loop, you can just do `q8`, no need to print
print(q8)
#  geom_point(aes(size= hp , cplour = hp ))
```
## Do a t-test


9.  What is the *p*-value for a *t*-test (using the mtcars data), that tests if
displacement (disp) varies between engine shapes (column "vs")?
```{r}
boxplot(mtcars$disp~mtcars$vs)
t.test(mtcars$disp ~ mtcars$vs, mu = 0, conf.level = 0.95, alternative = "two.sided")

#p-value = 2.477e-06
```


## Basic unix commands

10.  Which command is used to create a new directory?
```{bash, engine.opts='-l'}
mkdir test
```

11.  Which command shows the size of a file?
```{bash, engine.opts='-l'}
du -sh test
```

12.  Which command lets you delete a file?
```{bash, engine.opts='-l'}
rm test
```

13.  Which command copies a file?
```{bash, engine.opts='-l'}
 cp infoChallenge_2.fastq.gz aqk001/Scratch/matt/infoChallenge_2.fastq.gz 
```

14.  Which command moves a file from one directory to another?
```{bash, engine.opts='-l'}
mv infoChallenge_2.fastq.gz /aqk001/Scratch/matt/infoChallenge_2.fastq.gz
```

## Basic stats knowledge

15.  Does a *t*-test assume that the measurements follow a normal distribution?

A: Yes

16.  Briefly (1-2 sentences), what is the difference between parametric and
non-parametric tests?

A: Parametric tests make assumptions about the distribution of population from which the sample was taken. Non-parametric test does not make assumptions, that is, the data can be collected from a sample that does not follow a specific distribution (i.e normal distribution). Parametric test uses group means while Non-parametric analyses uses group medians

17.  Name one method to perform multiple testing correction

A:Bonferroni correction 
  Benjamini and Hochberg False Discovery Rate

18.  For a set of statistical tests, using a FDR cutoff of 0.1 will, in theory,
result in what percentage of false positives in your "significant" tests?

A: 10%

## Basic file/data formats

19.  What file/data format is this?

A: FASTA Format

```{verbatim, max.height='100px'}
>Sequence_1
AGTACATGGGGAGCTGGCGCAATGACGACCTAATTGGCGCACAGTACTAGGCATCGTCATCCAATGCGACGAGTCCTAC
ACTATCTTGGATATGATATGGCGCACTACACATGCTAGCCGCTGGGGAGATTAGCTCGAGTTGCCCCTTTGCCCGATCC
CGGAAGATACGCTCTAAGCTCGGCAATCGCTCTTGCCGTGCGAGATGCTAGCAAAAAGGTGTACTTCTCAGCGGAGCAG
AAAGATCATGTTTATTGGAAGCATCAACCTGCGCCGTCTTGTTAACTTGTCATATCGCGCACGTAGT
>Sequence_2
GCACTACACATGCTAGCCGCTGGGGAGATTAGCTCGAGTTGCCCCTTTGCCCGATCCCGGAAGATACGCTCTAAGCTCG
GCAATCGCTCTTGCCGTGCGAGATGCTAGCAA
```

20.  What file/data format is this?

A: FASTQ Format

```{verbatim, max.height='100px'}
@A00498:77:HGHFVDRXX:1:2101:3775:1016 1:N:0:NACGGCTC
ANTTAGGCAACCGTGCTTATGGAAATGCA
+
F#FFFFFFFFFFFFFFFFFFFFFFF:FF,
@A00498:77:HGHFVDRXX:1:2101:3830:1016 1:N:0:NATCGCTC
TNGCTGCAGAATAACCCGGGCTCAGACTC
+
F#FFFFFFFF,FFFFFFFFFFFFFFFFF:
```

21.  What file/data format is this?

A: BED file format

```{}
chr1  213941196  213942363
chr1  213942363  213943530
chr1  213943530  213944697
chr2  158364697  158365864
chr2  158365864  158367031
chr3  127477031  127478198
```

22.  What are the two main differences between a bam and sam file?

A: SAM files are human-readable text files, and BAM files are simply their binary equivalent.
BAM files are smaller and more efficient for software to work with than SAM files because BAM files are compressed.
------------------------------------------------------------------------

# Intermediate - 2 points per question

## Make fancy plots

23.  Using the mtcars data, make a plot of mpg vs disp, coloring the points by
cyl. Include a title. Be sure the colors used are color-blind friendly.

24.  From the previous plot, include a legend for point colors and make the
title of the legend "Cylinder".

25.  Add a linear regression line to the plot along with text in the plot 
providing the Pearson's correlation and the *p*-value
```{r}
#Answer for 23 to 25
library(ggpubr)
q23 <- ggplot(mtcars, aes(x=disp, y=mpg)) +
  geom_point(size=2, aes(colour = factor(cyl))) + 
  scale_color_brewer(palette="Dark2") +
  geom_smooth(method = lm, se= F) +
  ggtitle("Miles Per Gallon") +
    xlab("Displacement") + 
    ylab("Miles Per Gallon")+
  theme_bw()
q25 <- q23 %>% 
  + stat_cor( method="pearson", label.x = 310,label.y = 33)
print(q25)
```
## Align reads to a reference

26.  On the Franklin cluster, in the folder /gpfs0/scratch/mvc002/ are two files:

  -   infoChallenge_R1.fastq.gz
  -   infoChallenge_R2.fastq.gz
  
  Align these reads to the hg38 reference using your aligner of choice and share
  your code.
  Several pre-made references are available here:
  `/reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Sequence/`.
```{bash, engine.opts='-l'}
ml purge
module load HISAT2/2.2.1
gzip -d infoChallenge_1.fastq.gz
gzip -d infoChallenge_2.fastq.gz
hisat2 -x /reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Sequence/HiSat2Index/genome -1 infoChallenge_1.fastq -2 infoChallenge_2.fastq -S infoChallenge.bam

#MVC############################################################################
# be aware that hisat2 outputs sam format, not bam, so you're writing out a sam file named .bam, which isn't actually a bam file. To get a bam file get rid of the -S and pipe the output ` | samtools view -b - > infoChallenge.bam`
```

27.  Share a screenshot of some of the aligned reads in IGV

A: Will attach with email

28.  Is this dataset whole genome sequencing, RNAseq, ChIPseq or amplicon seq?

A: RNA seq

## --R-specific-- Demonstrate the use of pipes and tidyverse functions

29.  In R, starting with the mtcars dataset use `filter()` to remove any row
where cyl is "8" and then use `group_by()` to group the data by gear and then
use `summarize()` to calculate the average of the mpg column per group. This
should be in one continuous pipeline without intermediate variables or nested
functions. Use either pipe (`%>%` or `|>`).

```{r}
library(tidyverse)
#MVC############################################################################
# if you'd like, install lintr and run this document through it
# Try to avoid trailing whitespaces
mtcars %>% 
  filter(cyl != 8) %>% 
  group_by(gear) %>% 
  summarize(mpg)
#MVC############################################################################
# This just summarizes the tibble but doesn't calculate the mean per group
mtcars %>% 
  filter(cyl != 8) %>% 
  group_by(gear) %>% 
  summarize(gear_mean = mean(mpg))

```
## Write a bash loop

30.  Write a "for loop" in bash to loop over the numbers 1-10 and print them to
the console
```{bash, engine.opts='-l'}
for ((i=1; i<=10; i++)) do
echo "itration number is " $i
done
```
## Git and GitHub

31.  For your submission, put your results up into a public GitHub repository
and provide the link

A: Link:
https://github.com/aimanbarki/InfoChallenge

32.  Make sure your repo includes at least two branches and two commits

A: done

------------------------------------------------------------------------

# Hard - 3 points per question

## Harder plotting

33.  Add histograms to the top and right side of the mtcars mpg vs disp
scatterplot so that it looks like Figure 1. Note that the axes of each plot
should align.

![example plot]("C:\Users\mvc002\OneDrive - Nationwide Children's Hospital\Education\informatics_challenge\informatics_challenge_2022\exampleTriPlot.png")

https://nationwidechildrens-my.sharepoint.com/:i:/g/personal/matthew_cannon_nationwidechildrens_org/Ef7xSrUpIoJMsl2ua-KezHkBArzTwEkqwMNjLnZ8vMaZFQ?e=9kLFHD 

```{r}
library(ggExtra)
library(gridExtra)

pMain <- ggplot(mtcars, aes(x = mpg, y = disp)) +
         geom_point() + ylim(0, 480) + xlim(10, 35) + theme_linedraw()
pTop <- ggplot(mtcars, aes(x = mpg)) +  
        geom_histogram(show.legend = FALSE, bins = 120)+ xlim(10, 35) + theme_linedraw()
pRight <- ggplot(mtcars, aes(disp)) + xlim(0, 480) + 
        geom_histogram(show.legend = FALSE, bins = 110) + coord_flip() + theme_linedraw()
pEmpty <- ggplot(mtcars, aes(x = mpg, y = disp)) +
          geom_blank() +
          theme(axis.text = element_blank(),
                axis.title = element_blank(),
                line = element_blank(),
                panel.background = element_blank()) 

grid.arrange(pTop, pEmpty, pMain, pRight, ncol = 2, nrow = 2, widths = c(3, 1), heights = c(1, 3))   

#MVC############################################################################
# Almost! Good job using grid, but your x-axis isn't aligned, by default grid doesn't align the axes. You can use the cowplot package to do it
```
## Get sequences from SRA

34.  Using the command line, download the SRR10197556 dataset from the NCBI SRA
repository
```{bash, engine.opts='-l'}
ml SRAToolkit
prefetch -X 70G SRR10197556
fastq-dump --split-files --gzip SRR10197556/SRR10197556.sra
```
## Bash variable expansion

35.  Create a variable in bash containing the text "path/to/target/file.txt" and
then share code to trim off the path ("path/to/target/")
```{bash, engine.opts='-l'}
var=path/to/target/file.txt
echo $var | cut -d  "/" -f 4  
#MVC############################################################################
# I like the creative use of cut, another way to do it:
echo ${var##*/}
```
36.  Create a variable in bash containing the text "path/to/target/file.txt" and
then share code to trim off the file extension (".txt")

```{bash, engine.opts='-l'}
var=path/to/target/file.txt
echo $var | cut -d  "." -f 1  

#MVC############################################################################
# also good use of cut, another method:
echo ${var%.txt}
```
## Grep

37.  Use the Unix grep function to match the pattern "GATACA" in `/gpfs0/scratch/mvc002/infoChallenge_1.fastq.gz` on the Franklin cluster. How
many lines had a match?
```{bash, engine.opts='-l'}
gzip -d infoChallenge_1.fastq.gz
# i to make it case-insenstive
grep -ic GATACA infoChallenge_1.fastq
#MVC############################################################################
# this works, but will be slower since you had to unzip the file, I should have been more explicit in the question, but a better way is to use zgrep on the gzipped file

```
output number: 472703
## Write a program

38.  Share code written in a language other than R or bash that reads in a fastq
file and trims 10 characters from the 3' end of the read and quality score
fields. You can use the `/gpfs0/scratch/mvc002/infoChallenge_1.fastq.gz` file to
test your code. 

A: attaching q38_aiman.py

------------------------------------------------------------------------

# Too hard - 4 points per question

## Finding a sequence in a fasta file?

39.  How many times does sequence "GATTACA" appear in this fasta file? `/reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Sequence/WholeGenomeFasta/genome.fa`
```{bash, engine.opts='-l'}

grep -io GATTACA genome.fa | wc -l
#MVC############################################################################
# Almost! the one thing you didn't think about is that since the fasta file has the sequence line wrapped, so this approach will miss any pattern that starts on one line and wraps to the next. That's what makes this a harder question
```
## Using awk

40.  Use this gtf file (`/reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Annotation/Genes/genes.gtf`)
and the Unix `awk` command to keep only the lines where first column is "chr11"
and the fourth column is greater than 100,000. How many lines are in the output?
```{bash, engine.opts='-l'}
 awk '/chr11/ {if ($4 >= 100000) print}' genes.gtf |wc -l
 
 ###55736
#MVC###########################################################################
#Close, but your approach finds any line with "chr11" in it, which includes for instance "chr11_KI270927v1_alt"
  awk '$1 == "chr11" && $4 >= 100000 {print $_}' \
    /reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Annotation/Genes/genes.gtf \
    | wc -l
```
------------------------------------------------------------------------

# I hate myself - 5 point per question

## Identify the mystery data type

41.  On the Franklin cluster is a folder /gpfs0/scratch/mvc002/. Inside this
folder are two files:
  -   mysteryData_R1.fastq.gz
  -   mysteryData_R2.fastq.gz
  What kind of sequencing experiment are these data derived from?

42.  What species do these reads come from?
A: I converted these files to fasta and did the BLAST. It looks like the is mainly from human but some reads also came from Cyprinus carpio.

#MVC############################################################################
# Good try but not right - I'll explain on the 5th.

## Harder programming

43.  Share code that finds a specified sequence in a fasta file, which also
allows for ambiguous bases in the query
(https://droog.gs.washington.edu/parc/images/iupac.html)

A: attaching Q43_aiman.py

#MVC############################################################################
# I'm really impressed you were ablt to put this together without knowing python previously. Unfortunately, since you call print() three times, it's putting newlines after each one, so your output isn't in the right format (with each match as a row of three elements)
